### ISSUE TITLE
ecc/p384: ScalarMult fails for k=6

### ISSUE THREAD

**Issue by mmcloughlin** [cloudflare/circl#62]
The P384 `ScalarMult` implementation does not return the correct result for the scalar `k = 6`. For a reproducer see:

https://gist.github.com/mmcloughlin/823f698cd20cb299327bec705de0f014

One example result:

```
$ go run k6.go
got    x=0 y=0
expect x=78d0da4eeeb585f21e4de75c543807e8bd7d479a9b332f2cd5e4fbfc219e54d328ee02cce588805bf5ba9217fc52e1c8 y=723df279aaf436b7f088d60078cf26aef0410a51fbc1ec241532ebf321192749dfde358aef76e8874148c321761984d8
```

I discovered this because I noticed that the `ScalarMult` function does not implement the same algorithm as described in [Selecting Elliptic Curves for Cryptography: An Efficiency and Security Analysis](https://eprint.iacr.org/2014/130). It is very similar to Algorithm 1, but does not implement Step 19 which uses a complete addition formula for the final add.

![Screen Shot 2019-09-01 at 4 15 10 PM](https://user-images.githubusercontent.com/7133685/64083359-b4808c80-ccd3-11e9-9579-da94869f34a6.png)

The value k=6 has the property that the final add of the algorithm will be `(-3)P + (-3)P`. Therefore the `add` function fails, since it does not support doubling.

**Comment by armfazh** [comment]
[Renes-Costello-Batina](https://ia.cr/2015/1060)'s formulas were included in #66 for complete point addition, since they are required in the last iteration of main loop (Line 19, Alg. 1).

**Comment by armfazh** [comment]
[This script](https://gist.github.com/armfazh/b63f878b6ded15612c6b913a0d7959c2) looks for some values that trigger the issue.
Example: w=2 and k \in {6, -6}.
